Add kn-ps-harbor-slack function (#917)
vmware-samples/vcenter-event-broker-appliance: examples/knative/powershell/kn-ps-harbor-slack/README.md

# kn-ps-harbor-slack

> Example Knative PowerShell function for sending to a Slack webhook when the Harbor webhook feature is configured to send repository events to the [kn-go-harbor-webhook function example](https://github.com/vmware-samples/vcenter-event-broker-appliance/tree/development/examples/knative/go/kn-go-harbor-webhook).

Reviewer: nit: wording

```suggestion
Example Knative PowerShell function for sending Harbor CloudEvents to a Slack webhook. This function relies on the Harbor webhook [function example](https://github.com/vmware-samples/vcenter-event-broker-appliance/tree/development/examples/knative/go/kn-go-harbor-webhook) which is a requirement for this example.
```

Author: Done

# Step 1 - Build

> 

Reviewer: nit: remove newline

Author: Done

Create the container image locally to test your function logic.

```
export TAG=<version>

> docker build -t <docker-username>/kn-ps-harbor-slack:${TAG} .

Reviewer: btw: it's easier if you make this a var, e.g. IMAGE= which you can later reference in your README instead of always repeating that string.

Author: @embano1 Sure thing, but we consequently should consider that this is probably a new issue/PR for most of our function examples... See for example: https://github.com/vmware-samples/vcenter-event-broker-appliance/tree/development/examples/knative/powershell/kn-ps-slack-vsphere-alarm

Author: Like this: 

```shell
export IMAGE=<image>
export TAG=<version>
export DOCKERUSER=<docker id>
docker build -t ${DOCKERUSER}/${IMAGE}:${TAG} .
```

Author: Done

```

# Step 2 - Test

Verify the container image works by executing it locally.

Change into the `test` directory
```console
cd test
```

Update the following variable names within the `docker-test-env-variable` file

* SLACK_WEBHOOK_URL - Slack webhook URL
* SLACK_MESSAGE_PRETEXT - Text displayed for Slack notification

Start the container image by running the following command:

```console
docker run -e FUNCTION_DEBUG=true -e PORT=8080 --env-file docker-test-env-variable -it --rm -p 8080:8080 <docker-username>/kn-ps-harbor-slack:${TAG}
```

In a separate terminal, run either `send-cloudevent-test.ps1` (PowerShell Script) or `send-cloudevent-test.sh` (Bash Script) to simulate a CloudEvent payload being sent to the local container image

```console
Testing Function ...
See docker container console for output

# Output from docker container console

> Detected change to subject-123 ...

Reviewer: question: is this really an output from your function? "subject-123" doesn't seem like?

Author: It's not an output from my function. I simply reused the content from our [Slack function example](https://github.com/vmware-samples/vcenter-event-broker-appliance/tree/development/examples/knative/powershell/kn-ps-slack).

Reviewer: so it needs to be updated, right?

Author: Updated

                    Date: Tue, 16 Feb 2021 21:56:16 GMT
                    Server: Apache
                    Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
                    x-slack-backend: r
                    referrer-policy: no-referrer
                    Vary: Accept-…
Headers           : {[Date, System.String[]], [Server, System.String[]], [Strict-Transport-Security, System.String[]], [x-slack-backend, System.String[]]…}
Images            : {}
InputFields       : {}
Links             : {}
RawContentLength  : 2
RelationLink      : {}
```

# Step 3 - Deploy

> **Note:** The following steps assume a working Knative environment using the
`default` Rabbit `broker`. The Knative `service` and `trigger` will be installed in the
`vmware-functions` Kubernetes namespace, assuming that the `broker` is also available there.
>
> **Note:** Also, in order to receive incoming Harbor events and to ultimately invoke the Harbor-Slack-Function, it's necessary to have the [kn-go-harbor-webhook function example](https://github.com/vmware-samples/vcenter-event-broker-appliance/tree/development/examples/knative/go/kn-go-harbor-webhook) running properly first.

> Push your container image to an accessible registry such as Docker once you're done developing and testing your function logic.

Reviewer: This is not needed when using the provided function.yaml/image

Author: @embano1 Make sense. This would include line 69-71 as well, right?

Author: Done

```console
docker push <docker-username>/kn-ps-harbor-slack:${TAG}
```

Update the `slack_secret.json` file with your Slack webhook configurations and then create the kubernetes secret which can then be accessed from within the function by using the environment variable named called `SLACK_SECRET`.

```console
# create secret

kubectl -n vmware-functions create secret generic harbor-slack-secret --from-file=SLACK_SECRET=slack_secret.json

# update label for secret to show up in VEBA UI
kubectl -n vmware-functions label secret slack-secret app=veba-ui
```

> Edit the `function.yaml` file with the name of the container image from Step 1 if you made any changes. If not, the default VMware container image will suffice. By default, the function deployment will filter on the `com.vmware.harbor.push_artifact.v0` Harbor Event. If you wish to change this, update the `type` field within `function.yaml` to the desired event type.

Reviewer: nit: Here some help would be nice, e.g. a link to Harbor event notifications and/or how to inspect them, e.g. Sockeye

Author: Added:
A list of supported notification events is available on the official Harbor documentation under [Configure Webhook Notifications](https://goharbor.io/docs/2.5.0/working-with-projects/project-configuration/configure-webhooks/). Furthermore, use the VEBA Event viewer endpoint (`https://<VEBA-FQDN>/events`) to display all incoming events.

Author: Done

> 

Reviewer: nit: remove

Author: Done
